### PR TITLE
chore: update to Docker Compose V2

### LINKED ISSUE #13938: Upgrade docker compose to V2
# General

[Docker compose V2 became general available in April 2022](https://www.docker.com/blog/announcing-compose-v2-general-availability/). With that general availability [a deprecation plan for docker compose V1 was given](https://www.docker.com/blog/announcing-compose-v2-general-availability/#proposed-compose-v1-end-of-lie-timeline). This plan essentially states that in October 2022 the Support of critical bug fixes and severe security issues will end on Compose v1. Thus docker compose should be updated to V2.

In addition, with regard to the ongoing containerisation of components, the additional features of docker compose V2 can be very helpful and avoid a lot of manual workarounds and mitigations for unavailable functionality in V1.

# Implementation details

- [x] Install docker compose V2 on all necessary VMs and for all necessary deployments 
  - [x] Magma Dev VM (containerized AGW, FeG)
  - [x] CWAG VM (?)
  - [x] Developer laptop for orc8r
- [x] Adopt any usage of docker compose within the code
- [ ] Replace workarounds for docker compose V1 shortcomings with proper solutions
- [x] Update any documentation to use docker compose V2
- [x] Remove docker compose V1 from the Magma Dev VM -> with new base image
- [ ] Remove workaround from https://github.com/magma/magma/pull/14090 (:warning: The containerized LTE integ test action does not work on master.)
- [x] Consider releasing new base images with the docker-compose upgrade

# Hints

- There is a script or tool for the migration of Docker compose from V1 to V2.

# Open points

- Is docker compose needed on any other VM, but the Dev VM?

### PR DESCRIPTION
**Pull request by mpfirrmann**
Docker is  [deprecating docker compose V1](https://www.docker.com/blog/announcing-compose-v2-general-availability/#proposed-compose-v1-end-of-lie-timeline).
Additionally, the new features of docker compose V2 may avoid a lot of manual workarounds and mitigations for unavailable functionality in V1.

Linked to #13938.

:warning: There would be a chicken and egg problem in CI with the naming changes introduced by Docker Compose V2 . Our (post merge) CI workflows build and upload Docker images to the artifactory. These built images are then used to perform various integration tests. The GitHub actions that run on PRs download the latest images to test a subset of these workflows against the PR branch. In order to keep the required jobs happy, two PRs would be needed to change the naming scheme of these images on the artifactory. The first one changing the upload destination and a second one changing the download source. :warning: 
This PR does not rename the packages on the artifactory to avoid this problem for now.

## Test Plan

- [x] Build Orc8r, SwaggerAPI, NMS along the [Quick Start Guide](https://magma.github.io/magma/docs/next/basics/quick_start_guide)
- [x] Build containerized AGW along the (adapted) [README.md](https://github.com/magma/magma/blob/6ff47d2d37a406c318fcb90c4d6a9f965ef11019/lte/gateway/docker/README.md)
- [x] LTE integration tests (incl. traffic server tests) w/ containerized AGW, [precommit](https://github.com/mpfirrmann/magma/actions/runs/3729581414) and [extended](https://github.com/mpfirrmann/magma/actions/runs/3729597236)
- [x] [CWAG build and integration test suite](https://github.com/mpfirrmann/magma/actions/runs/3730672081)
- [x] [FEG](https://github.com/mpfirrmann/magma/actions/runs/3729555430)
- [x] Dev Container
- Most GitHub Actions / CI only to be seen in CI / [Artifactory](https://linuxfoundation.jfrog.io/ui/repos/tree/General/magma-docker)

## Additional Information

- The Dev Container uses [Moby](https://mobyproject.org/) by default, instead of Docker. The latest `moby-compose` package includes Docker Compose V2.
- GitHub's `ubuntu-2004` action runners already included Docker Compose V2.12.2 (see [here](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md#tools)), commands were adapted. This can be seen for example in [this log](https://github.com/mpfirrmann/magma/actions/runs/3703637666/jobs/6275320075#step:4:3) of the FEG integ test run.
- Docker Compose V2 introduced a braking change, using `-` (dash) instead of `_` (underscore) as a naming prefix. This PR adapts the new naming convention via the `DOCKER_REGISTRY` variables.
- The formerly used `--parallel` flag was removed since Docker Compose V2 builds in parallel by default.

### CODE REVIEW

File: cwf/gateway/helm/cwf-virtlet/templates/bin/_install_gateway.sh.tpl
Comment (wolfseb) [inline cwf/gateway/helm/cwf-virtlet/templates/bin/_install_gateway.sh.tpl]: typo: missing 'n'

File: feg/gateway/helm/feg-kubevirt/templates/bin/_install_gateway.sh.tpl
Comment (wolfseb) [inline feg/gateway/helm/feg-kubevirt/templates/bin/_install_gateway.sh.tpl]: typo: missing 'n'

File: orc8r/tools/docker/install_gateway.sh
Comment (wolfseb) [inline orc8r/tools/docker/install_gateway.sh]: nit: I suggest alphabetically sorted installs

**Comment by crasu** [comment]
@mpfirrmann 
* Has anyone except you tried this?
* Did you run the feg integ tests?

**Comment by wolfseb** [comment]
>     * Now that the image tags are set in the yml files (e.g. "cwf_nginx"), can they be removed somewhere else? Or were they set automatically before?

That's what the `COMPOSE_PROJECT_NAME` variable did (it is prepended to the image names unless overwritten by explicitly specifying an image tag).

**Review by wolfseb — APPROVED** [review]

File: feg/gateway/helm/feg-virtlet/templates/bin/_install_gateway.sh.tpl
Comment (MoritzThomasHuebner) [inline feg/gateway/helm/feg-virtlet/templates/bin/_install_gateway.sh.tpl:159]: Line 158 still needs to be changed to `docker compose pull`

**Comment by MoritzThomasHuebner** [comment]
I found two more spots where I believe you need to change a `docker-compose` to a `docker compose`:

`lte/gateway/deploy/agw_install_docker.sh` Line 70
`docs/readmes/cwf/deploy_install.md` Line 226 (This one is just a piece of documentation)

**Comment by mpfirrmann** [comment]
>     * You removed the COMPOSE_PROJECT_NAME variables in the environment settings but not for NMS (magmalte). Any reason why?
 
Just tested this, can be removed as well.

File: docs/readmes/cwf/deploy_install.md
Comment (voisey) [inline docs/readmes/cwf/deploy_install.md:138]: Looks like you might have missed one on line 226 in this file

File: orc8r/cloud/docker/.env
Comment (voisey) [inline orc8r/cloud/docker/.env:12]: Does removing these variables have unintended side effects? E.g. something like this could be affected https://github.com/magma/magma/blob/master/orc8r/tools/docker/publish.sh#L50

**Comment by mpfirrmann** [comment]
> I found two more spots where I believe you need to change a `docker-compose` to a `docker compose`:
> 
> `lte/gateway/deploy/agw_install_docker.sh` Line 70 
I adapted this and tested it with a bare ubuntu machine following the [Install Docker AGW docs](https://magma.github.io/magma/docs/lte/deploy_install_docker).

> `docs/readmes/cwf/deploy_install.md` Line 226 (This one is just a piece of documentation)
Changed

File: cwf/gateway/helm/cwf-kubevirt/templates/bin/_install_gateway.sh.tpl
Comment (sebathomas) [inline cwf/gateway/helm/cwf-kubevirt/templates/bin/_install_gateway.sh.tpl:104]: Why?

Comment (mpfirrmann) [reply cwf/gateway/helm/cwf-kubevirt/templates/bin/_install_gateway.sh.tpl:104]: Changed. I initially thought, we don't need it. But `add-apt-repository` needs `gnupg`.

**Comment by mpfirrmann** [comment]
>     * Did you run the feg integ tests?

Yes, they did run green locally and in [CI on my fork](https://github.com/mpfirrmann/magma/actions/runs/3539845460).

File: orc8r/cloud/docker/.env
Comment (mpfirrmann) [reply orc8r/cloud/docker/.env:12]: The `COMPOSE_PROJECT_NAME` is used as a prefix to publish FEG and CWF docker images. Instead of deleting the variables and letting docker do its naming magic, I used the (at some parts already used) `DOCKER_REGISTRY` variable to mitigate the switch to the "new" naming convention. The publishing script now still has it's prefix variable.

File: .github/workflows/federated-integ-test.yml
Comment (voisey) [inline .github/workflows/federated-integ-test.yml]: Here you change the tarred files to use a dash rather than an underscore, but for the orc8r components above you don't. Is there a reason for this?

**Comment by sebathomas** [comment]
What about the latest build error?
https://github.com/magma/magma/actions/runs/3583662473/jobs/6030089425

File: .github/workflows/federated-integ-test.yml
Comment (mpfirrmann) [reply .github/workflows/federated-integ-test.yml]: Let's do the naming of images on the LF artifactory once the switch is done. This is also the reason for the failing test mentioned by @sebathomas.
I'm not sure if we can leave the externally visible package names as they are (with an underscore), or if everything has to be changed to a dash naming scheme.

**Comment by MoritzThomasHuebner** [comment]
CWAG integ tests are still red (probably due to flakiness). Maybe try re-running.

File: .github/workflows/federated-integ-test.yml
Comment (mpfirrmann) [reply .github/workflows/federated-integ-test.yml]: I decided to leave the naming scheme on the artifactory as is. We would need two PRs for this anyway and it would only add up to the complexity of the PR.

File: cwf/gateway/helm/cwf-kubevirt/README.md
Comment (MoritzThomasHuebner) [inline cwf/gateway/helm/cwf-kubevirt/README.md:78]: Line 12 of this file: `docker_registry: docker.io/cwf_`. Should the "_" be replaced with a "-"? The same prefix is also still in helm/cwf-kubevirt/values.yaml and  helm/cwf-virtlet/README.md.

File: .github/workflows/build_all.yml
Comment (MoritzThomasHuebner) [inline .github/workflows/build_all.yml:314]: In line 324 you can remove the `--parallel` flag.

**Comment by MoritzThomasHuebner** [comment]
https://github.com/magma/magma/pull/14415/files#diff-8a35a67f518ee5aa3cf2db81e814916f691acbc335c9a8b851b5cf0836e9834aL165

Does this need to be changed to `ssh testuser@${feg_ip} "/var/opt/magma/docker/docker compose exec magmad /usr/local/bin/show_gateway_info.py"` (replace hyphen with space)?

**Comment by MoritzThomasHuebner** [comment]
FEG and CWF look good. Some CWF integ test pass on retry but this is normal. See my other comments, they should be relatively trivial. I'm already approving so I don't forget it before the holidays.

File: cwf/gateway/helm/cwf-kubevirt/README.md
Comment (mpfirrmann) [reply cwf/gateway/helm/cwf-kubevirt/README.md:78]: The uploaded packages keep their names, so this is correct.

**Comment by mpfirrmann** [comment]
> https://github.com/magma/magma/pull/14415/files#diff-8a35a67f518ee5aa3cf2db81e814916f691acbc335c9a8b851b5cf0836e9834aL165
> 
> Does this need to be changed to `ssh testuser@${feg_ip} "/var/opt/magma/docker/docker compose exec magmad /usr/local/bin/show_gateway_info.py"` (replace hyphen with space)?

I adapted it to "cd /var/opt/magma/docker/; docker compose exec magmad /usr/local/bin/show_gateway_info.py".

**Review by sebathomas — APPROVED** [review]
I tried some builds on my machine and didn't notice any errors.

**Review by MoritzThomasHuebner — APPROVED** [review]

**Review by sklgromek — APPROVED** [review]

**Review by nstng — APPROVED** [review]
went over changes in pairing - lgtm, but should be monitored after merge and checked if all images are created correctly

**Pull request merged**
